core/src/block_creation_loop.rs

```diff
@@ -0,0 +1,521 @@
+            // Give up on this leader window
+            error!(
+                "{my_pubkey}: Unable to produce window {start_slot}-{end_slot}, skipping window: \
+                 {e:?}"
+            );
+            continue;
+        }
+    }
+
+    info!("{my_pubkey}: Block creation loop shutting down");
+}
+
+/// Resets poh recorder
+fn reset_poh_recorder(bank: &Arc<Bank>, ctx: &LeaderContext) {
+    trace!("{}: resetting poh to {}", ctx.my_pubkey, bank.slot());
+    assert!(ctx.record_receiver.is_shutdown() && ctx.record_receiver.is_safe_to_restart());
+    let next_leader_slot = ctx.leader_schedule_cache.next_leader_slot(
+        &ctx.my_pubkey,
+        bank.slot(),
+        bank,
+        Some(ctx.blockstore.as_ref()),
+        GRACE_TICKS_FACTOR * MAX_GRACE_SLOTS,
+    );
+
+    ctx.poh_recorder
+        .write()
+        .unwrap()
+        .reset(bank.clone(), next_leader_slot);
+}
+
+/// Produces the leader window from `start_slot` -> `end_slot` using parent
+/// `parent_slot` while abiding to the `skip_timer`
+fn produce_window(
+    start_slot: Slot,
+    end_slot: Slot,
+    parent_slot: Slot,
+    skip_timer: Instant,
+    ctx: &mut LeaderContext,
+) -> Result<(), StartLeaderError> {
+    // Insert the first bank
+    start_leader_retry_replay(start_slot, parent_slot, skip_timer, ctx)?;
+
+    let my_pubkey = ctx.my_pubkey;
+    let mut slot = start_slot;
+    while !ctx.exit.load(Ordering::Relaxed) {
+        let leader_index = leader_slot_index(slot);
+        let timeout = block_timeout(leader_index);
+        trace!(
+            "{my_pubkey}: waiting for leader bank {slot} to finish, remaining time: {}",
+            timeout.saturating_sub(skip_timer.elapsed()).as_millis(),
+        );
+
+        if let Err(e) = record_and_complete_block(
+            ctx.poh_recorder.as_ref(),
+            &mut ctx.record_receiver,
+            skip_timer,
+            timeout,
+        ) {
+            panic!("PohRecorder record failed: {e:?}");
+        }
+        assert!(!ctx.poh_recorder.read().unwrap().has_bank());
+
+        // Produce our next slot
+        slot += 1;
+        if slot > end_slot {
+            trace!("{my_pubkey}: finished leader window {start_slot}-{end_slot}");
+            break;
+        }
+
+        // Although `slot - 1`has been cleared from `poh_recorder`, it might not have finished processing in
+        // `replay_stage`, which is why we use `start_leader_retry_replay`
+        start_leader_retry_replay(slot, slot - 1, skip_timer, ctx)?;
+    }
+
+    Ok(())
+}
+
+/// Records incoming transactions until we reach the block timeout.
+/// Afterwards:
+/// - Shutdown the record receiver
+/// - Clear any inflight records
+/// - Insert the alpentick
+/// - Clear the working bank
+fn record_and_complete_block(
+    poh_recorder: &RwLock<PohRecorder>,
+    record_receiver: &mut RecordReceiver,
+    block_timer: Instant,
+    block_timeout: Duration,
+) -> Result<(), PohRecorderError> {
+    loop {
```

Comment:
nit: could be `while !block_timeout.saturating_sub(block_timer.elapsed()).is_zero() {`

Fewer lines, but not sure which is more readable... could also add a comment above the loop like `// loop until slot is over`